src/bridge/MatrixBanSync.ts

```diff
@@ -0,0 +1,143 @@
+        case "org.matrix.mjolnir.rule.user":
+            return BanEntityType.User;
+        case "m.policy.rule.server":
+        case "org.matrix.mjolnir.rule.server":
+            return BanEntityType.Server
+        default:
+            return null;
+    }
+}
+
+const supportedRecommendations = [
+    "org.matrix.mjolnir.ban", // Used historically.
+    "m.ban"
+];
+
+export class MatrixBanSync {
+    private bannedEntites = new Map<string, BanEntity>();
+    private subscribedRooms = new Set<string>();
+    constructor(private config: MatrixBanSyncConfig) { }
+
+    public async syncRules(intent: Intent) {
+        this.bannedEntites.clear();
+        this.subscribedRooms.clear();
+        for (const roomIdOrAlias of this.config.rooms) {
+            try {
+                const roomId = await intent.join(roomIdOrAlias);
+                this.subscribedRooms.add(roomId);
+                const roomState = await intent.roomState(roomId, false) as WeakStateEvent[];
+                for (const evt of roomState) {
+                    this.handleIncomingState(evt, roomId);
+                }
+            }
+            catch (ex) {
+                log.error(`Failed to read ban list from ${roomIdOrAlias}`, ex);
+            }
+        }
+    }
+
+    /**
+     * Is the given room considered part of the bridge's ban list set.
+     * @param roomId A Matrix room ID.
+     * @returns true if state should be handled from the room, false otherwise.
+     */
+    public isTrackingRoomState(roomId: string): boolean {
+        return this.subscribedRooms.has(roomId);
+    }
+
+    /**
+     * Checks to see if the incoming state is a recommendation entry.
+     * @param evt A Matrix state event. Unknown state events will be filtered out.
+     * @param roomId The Matrix roomID where the event came from.
+     * @returns `true` if the event was a new ban, and existing clients should be checked. `false` otherwise.
+     */
+    public handleIncomingState(evt: WeakStateEvent, roomId: string) {
+        const content = evt.content as unknown as MPolicyContent;
+        const entityType = eventTypeToBanEntityType(evt.type);
+        if (!entityType) {
+            return false;
+        }
+        const key = `${roomId}:${evt.state_key}`;
+        if (evt.content.entity === undefined) {
+            // Empty, delete instead.
+            log.info(`Deleted ban rule ${evt.type}/$ matching ${key}`);
+            this.bannedEntites.delete(key);
+            return false;
+        }
+        if (!supportedRecommendations.includes(content.recommendation)) {
+            return false;
+        }
+        if (typeof content.entity !== "string" || content.entity === "") {
+            throw Error('`entity` key is not valid, must be a non-empty string');
+        }
+        this.bannedEntites.set(key, {
+            matcher: new MatrixGlob(content.entity),
+            entityType,
+            reason: content.reason || "No reason given",
```

Comment:
Could be a number, an object, `true` or other(?) valid `content` value sub type.
But I don't think that actually breaks anything..
Worst case, your message could be that you were banned because of `[Object object]`.

Reply:
Yes, your admin is very silly if they put silly reasons in, I think.